File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/MappedList.java
package com.cedarsoft.commons.javafx;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;
import javafx.collections.transformation.TransformationList;

/**
 * Maps elements of type IN to OUT with change listeners working as expected.
 */
public class MappedList<IN, OUT> extends TransformationList<OUT, IN> {
  private final Function<IN, OUT> mapper;
  private final ArrayList<OUT> mapped;

  /**
   * Creates a new MappedList list wrapped around the source list.
   * Each element will have the given function applied to it, such that the list is cast through the mapper.
   */
  public MappedList(ObservableList<? extends IN> source, Function<IN, OUT> mapper) {
    super(source);
    this.mapper = mapper;
    this.mapped = new ArrayList<>(source.size());
    mapAll();
  }

  private void mapAll() {
    mapped.clear();
    for (IN val : getSource()) {
      mapped.add(mapper.apply(val));
    }
  }

  @Override
  protected void sourceChanged(ListChangeListener.Change<? extends IN> c) {
    // Is all this stuff right for every case? Probably it doesn't matter for this app.
    beginChange();
    while (c.next()) {
      if (c.wasPermutated()) {
        int[] perm = new int[c.getTo() - c.getFrom()];
        for (int i = c.getFrom(); i < c.getTo(); i++) {
          perm[i - c.getFrom()] = c.getPermutation(i);
        }
        nextPermutation(c.getFrom(), c.getTo(), perm);
      }
      else if (c.wasUpdated()) {
        for (int i = c.getFrom(); i < c.getTo(); i++) {
          remapIndex(i);
          nextUpdate(i);
        }
      }
      else {
        if (c.wasRemoved()) {
          // Removed should come first to properly handle replacements, then add.
          List<OUT> removed = mapped.subList(c.getFrom(), c.getFrom() + c.getRemovedSize());
          List<OUT> duped = new ArrayList<>(removed);
          removed.clear();
          nextRemove(c.getFrom(), duped);
        }
        if (c.wasAdded()) {
          for (int i = c.getFrom(); i < c.getTo(); i++) {
            mapped.addAll(c.getFrom(), c.getAddedSubList().stream().map(mapper).collect(Collectors.toList()));
            remapIndex(i);
          }
          nextAdd(c.getFrom(), c.getTo());
        }
      }
    }
    endChange();
  }

  private void remapIndex(int i) {
    if (i >= mapped.size()) {
      for (int j = mapped.size(); j <= i; j++) {
        mapped.add(mapper.apply(getSource().get(j)));
      }
    }
    mapped.set(i, mapper.apply(getSource().get(i)));
  }

  @Override
  public int getSourceIndex(int index) {
    return index;
  }

  /**
   * Do not @Override annotation since this method has only been added in Java 9
   */
  public int getViewIndex(int index) {
    return index;
  }

  @Override
  public OUT get(int index) {
    return mapped.get(index);
  }

  @Override
  public int size() {
    return mapped.size();
  }
}

File: commons/javafx/src/main/kotlin/com/cedarsoft/commons/javafx/MappedList.kt
package com.cedarsoft.commons.javafx

import javafx.collections.ListChangeListener
import javafx.collections.ObservableList
import javafx.collections.transformation.TransformationList
import java.util.function.Function
import java.util.stream.Collectors

/**
 * Maps elements of type IN to OUT with change listeners working as expected.
 */
class MappedList<IN, OUT>(
  source: ObservableList<out IN>,
  mapper: Function<IN, OUT>,
) : TransformationList<OUT, IN>(source) {
  private val mapper: Function<IN, OUT>
  private val mapped: ArrayList<OUT>

  /**
   * Creates a new MappedList list wrapped around the source list.
   * Each element will have the given function applied to it, such that the list is cast through the mapper.
   */
  init {
    this.mapper = mapper
    mapped = ArrayList(source.size)
    mapAll()
  }

  private fun mapAll() {
    mapped.clear()
    for (currentValue in source) {
      mapped.add(mapper.apply(currentValue))
    }
  }

  override fun sourceChanged(c: ListChangeListener.Change<out IN>) {
    // Is all this stuff right for every case? Probably it doesn't matter for this app.
    beginChange()
    while (c.next()) {
      if (c.wasPermutated()) {
        val perm = IntArray(c.to - c.from)
        for (i in c.from until c.to) {
          perm[i - c.from] = c.getPermutation(i)
        }
        nextPermutation(c.from, c.to, perm)
      } else if (c.wasUpdated()) {
        for (i in c.from until c.to) {
          remapIndex(i)
          nextUpdate(i)
        }
      } else {
        if (c.wasRemoved()) {
          // Removed should come first to properly handle replacements, then add.
          val removed = mapped.subList(c.from, c.from + c.removedSize)
          val duped: List<OUT> = ArrayList(removed)
          removed.clear()
          nextRemove(c.from, duped)
        }
        if (c.wasAdded()) {
          for (i in c.from until c.to) {
            mapped.addAll(c.from, c.addedSubList.stream().map(mapper).collect(Collectors.toList()))
            remapIndex(i)
          }
          nextAdd(c.from, c.to)
        }
      }
    }
    endChange()
  }

  private fun remapIndex(i: Int) {
    if (i >= mapped.size) {
      for (j in mapped.size..i) {
        mapped.add(mapper.apply(source[j]))
      }
    }
    mapped[i] = mapper.apply(source[i])
  }

  override fun getSourceIndex(index: Int): Int {
    return index
  }

  /**
   * Do not @Override annotation since this method has only been added in Java 9
   */
  fun getViewIndex(index: Int): Int {
    return index
  }

  override fun get(index: Int): OUT {
    return mapped[index]
  }

  override val size: Int
    get() {
      return mapped.size
    }
}
